jni_bindings/src/lib.rs

```diff
@@ -278,176 +341,192 @@ fn from_attr(attr: &dom::Attribute) -> Attribute {
 }
 
 #[no_mangle]
-pub extern "system" fn Java_org_phoenixframework_liveview_lib_Document_get_1node_1leaf_1string<
+pub unsafe extern "system" fn Java_org_phoenixframework_liveview_lib_Document_get_1node_1leaf_1string<
     'local,
 >(
-    env: JNIEnv<'local>,
+    mut env: JNIEnv<'local>,
     _: JClass<'local>,
+    // non-null raw pointer to Document
     this: jlong,
+    // NodeRef is the distinct u32 key mapped to a Node (and not a Node raw pointer)
     node_ref: jint,
 ) -> JString<'local> {
-    let mut clone = unsafe { env.unsafe_clone() };
-    let result = catch_unwind(|| {
-        let node = u32::try_from(node_ref).expect("value beyond `u32` range");
-        let node = NodeRef::new(node as usize);
-        let doc = unsafe { jlong_to_pointer::<dom::Document>(this).as_ref().unwrap() };
+    let this = this as *mut Document;
+
+    if this.is_null() {
+        env.throw_new(
+            "java/lang/NullPointerException",
+            "Document::get_node_leaf_string called with `this` as null pointer",
+        )
+        .unwrap();
+
+        JObject::null().into()
+    } else {
+        let node = NodeRef::new(node_ref as usize);
+        let doc = &*this;
+
         match doc.get(node) {
-            dom::Node::Leaf(ref s) => Ok(from_std_string_jstring(s.to_string(), env)),
-            _ => Err(String::from("node isn't a leaf")),
-        }
-    });
-
-    match result {
-        Ok(jstring) => jstring.unwrap(),
-        Err(err) => {
-            let exception_class = clone.find_class("java/lang/RuntimeException").unwrap();
-            let message = format!("Rust panic occurred in get_node_leaf_string: {:?}", err);
-            clone.throw_new(exception_class, message).unwrap();
-            //convert message to jstring and return
-            JObject::null().into()
+            dom::Node::Leaf(ref s) => env
+                .new_string(s.as_str())
+                .unwrap_or_else(|_| JObject::null().into()),
+            _ => JObject::null().into(),
         }
     }
 }
 
 #[no_mangle]
-pub extern "system" fn Java_org_phoenixframework_liveview_lib_Document_get_1node_1type(
+pub unsafe extern "system" fn Java_org_phoenixframework_liveview_lib_Document_get_1node_1type(
     mut env: JNIEnv,
     _: JClass,
+    // non-null raw pointer to Node
```

Comment:
The `non-null` can't be true or there would be no need for the `this.is_null()` check below.
```suggestion
    // raw pointer to Node
```